02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,267 @@
+
+In general, calling a method on an object like `.concat()` or `.push()` will change the underlying object, while any operation that contains an `=` will result in reassignment.
+
+### Passing Parameters
+
+**Question:** When we pass a parameter to a method, what do you get?
+- Is it the same underlying object?
+- Is it the same variable?
+- How can we find out?
+
+Let's write some code that will help us investigate this.
+
+```ruby
+def reassign_parameter(param)
+  puts "  Inside reassign_parameter"
+  puts "  at start, param.object_id is #{param.object_id}"
+
+  # .push modifies the underlying object
+  param.push('gecko')
+  puts "  after modification, param.object_id is #{param.object_id}"
+
+  # = changes the reference
+  param = ["new", "array"]
+  puts "  after reassignment, param.object_id is #{param.object_id}"
+  puts "  with value #{param}"
+  puts "  Finish reassign_parameter"
+end
+
+pets = ["dog", "parrot", "cat", "llama"]
+puts "Before reassign_parameter"
+puts "pets.object_id is #{pets.object_id}"
+puts
+
+reassign_parameter(pets)
+
+puts
+puts "After reassign_parameter"
+puts "pets.object_id is #{pets.object_id}"
+puts "with value #{pets}"
+```
+
+Before running this code, take a couple minutes to read through it. What is it doing? What do you expect the output to be?
+
+Running the code yields (your `object_id`s may be different):
+
+```
+Before reassign_parameter
+pets.object_id is 70144030241620
+
+  Inside reassign_parameter
+  at start, param.object_id is 70144030241620
+  after modification, param.object_id is 70144030241620
+  after reassignment, param.object_id is 70144030228060
+  with value ["new", "array"]
+  Finish reassign_parameter
+
+After reassign_parameter
+pets.object_id is 70144030241620
+with value ["dog", "parrot", "cat", "llama", "gecko"]
+```
+
+We can make a few interesting observations about this output:
+- The parameter inside the method has the same `object_id` as the variable we passed from outside
+- Modifications to the underlying object are visible outside the method
+- Reassigning the parameter with `=` does not reassign the outer variable
+
+This is exactly the same behavior we saw before, when we had two variables referencing the same object. From this we can conclude: **when you pass a variable as parameter, Ruby creates a new variable that references same object**.
+
+## Fixing the `short_strings` Method
+
+**Question:** Given what we've learned, how can we modify our `short_strings` method to do what we want?
+
+The answer is to modify the underlying object, rather than reassigning the parameter. Here's what the resulting code might look like:
+
+```ruby
+def short_strings_2(input)
+  input.each_with_index do |word, i|
+    # Slice characters 0 to 2
+    input[i] = word[0..2]
```

Comment:
Hmmm... You're right. I was focusing on the return of `each_with_index` being the original array  and not the fact that the block is mutating it. Can you add the ruby comments you added to this comment to the textbook code?